### ISSUE TITLE
Cannot deploy/remove stack because S3 bucket is gone

### ISSUE THREAD

**Issue by sbstjn** [serverless/serverless#3964]
I just ran into a situation where `sls remove` (1.17.0) failed to remove my service, because a CloudFormation resource was still updated by Amazon and the stack could not be removed. This totally makes sense, but my problem is that serverless/CloudFormation already deleted everything from S3 and now I'm stuck with a service that neither can be `deployed` nor `removed` because the S3 bucket does not exist anymore.

I'm fine if this is just based on my weird edge-case or on the custom CloudFormation resources and there is nothing you can do to prevent this. But a more specific error message than `The specified bucket does not exist` might be useful ;)

After deleting the Stack manually using the AWS Console I can deploy the service again. Maybe this is something that could be added doing with `sls remove --force` ?

## Console Outputs

```bash
$ > $(npm bin)/sls deploy
…
Serverless: Checking Stack update progress...
Serverless: Stack update finished...
```

```bash
$ > $(npm bin)/sls remove
Serverless: Getting all objects in S3 bucket...
Serverless: Removing objects in S3 bucket...
Serverless: Removing Stack...
Serverless: Checking Stack removal progress...
Serverless: Deployment failed!
 
  Serverless Error ---------------------------------------
 
  An error occurred while provisioning your stack: CustomTableA - Attempt to change a resource which is still in use: Table: custom-table-A is in the process of being updated..
```

### State where neither deploy nor remove work anymore

```bash
$ > $(npm bin)/sls remove
Serverless: Getting all objects in S3 bucket...
 
  Serverless Error ---------------------------------------
 
  The specified bucket does not exist
```

```bash
$ > $(npm bin)/sls deploy
Serverless: Packaging service...

  Serverless Error ---------------------------------------
 
  The specified bucket does not exist
```
--- 

#### Solution implementation proposal

_(Added by maintainers)_

Ensure that _bucket does not exist_ error does not trigger exception. Instead removal process should continue with stack removal

It can be fixed here: https://github.com/serverless/serverless/blob/c864fbd4826de27d2796e394b0a100c8d3add33e/lib/plugins/aws/remove/lib/bucket.js#L52-L53
If either `this.listObjects` or `this.deleteObjects` fails with _bucket does not exist error_. ignore such error silently

It might be that it's not sufficient for some cases. Ideally then we should force removal of a stack and eventually list all resources that could not have been removed

**Comment by horike37** [comment]
IMO, it would chage the error message something like this: `The specified bucket has been deleted. Please remove your stack with sls remove --force`.
`--force` option has not been added yet, we should add this to sls core though.

**Comment by pmuens** [comment]
Thanks for opening @sbstjn 👍

Yes, that's definitely an annoying issue (I too stumbled upon this one a couple of times).

Great proposal @horike37 👍

BTW the `--force` flag usage / stack removal enforcement is discussed in https://github.com/serverless/serverless/issues/3928.

**Comment by laurenfb** [comment]
+1 on this. I was hoping `sls remove` + `sls deploy` would fix some weird issues I've been having, but now that I've `remove`d I can't redeploy since the bucket doesn't exist. Is there a way around this without needing to create an entirely new instance? Recreate the bucket by hand?

**Comment by pmuens** [comment]
Thanks for the comment @laurenfb 👍

> Is there a way around this without needing to create an entirely new instance? Recreate the bucket by hand?

Have you tried to define the `ServerlessDeploymentBucket` in the [`resources` section](https://serverless.com/framework/docs/providers/aws/guide/resources/) in `serverless.yml`?

**Comment by TimCoates** [comment]
Also getting stuck with the "The specified bucket does not exist" problem. Something somewhere obviously expects a specific bucket to exist, where is that, so I can either create the bucket or remove the reference please?

**Comment by pmuens** [comment]
Thanks for commenting @TimCoates 👍

The resource for the bucket will be automatically created by Serverless and is called `ServerlessDeploymentBucket`. You can [manipulate this resource](https://serverless.com/framework/docs/providers/aws/guide/resources/) and overwrite it with something custom if you want to.

**Comment by iaforek** [comment]
I had the same problem. Yesterday, I modified a Role through AWS Console and when I did `sls remove` something didn't go according to the plan and I've got stuck with: `The specified bucket does not exist`. Neither `sls remove` or `sls deploy` were working. In the end I went back to AWS Console and in Cloud Formation I found an associated stack with status "DELETE_FAILED" - manually deleted it and now I can redeploy. Moreover, you can try to go to S3 and look for `ServerlessDeploymentBucket` and delete that too.

**Comment by albertoantunes** [comment]
Ran into this issue today as well. Doing as @iaforek comment (manually delete CloudFormation stack in AWS console) resolved the issue. In my case the problem is that we had manually attached a policy to the Role created by serverless and when we tried to execute serverless remove that caused it to error and be in this stuck state.

**Comment by rsweetland** [comment]
I just ran into this as well. A few unmentioned tips to get things reset: 

 * The AWS Console CloudFormation "delete" action may refuse to delete your stack the first time. If this happens you (of course) just try to delete it again. The second time it gives you an option to skip the offending resource. In my case, it was an IAM role. I proceeded with the delete, then deleted the role. 
 * You may have to delete the CloudWatch log group, too. Fortunately the Serverless CLI outputs this group which can be searched for in the CloudWatch UI.

I skimmed some of the discussion on the on the ```--force ``` flag. To make this slightly safer, maybe it could ask for user confirmation of each resource it is force-deleting as it goes? Having the user go directly to AWS is a leaky abstraction of Serverless magic.

**Comment by wviana** [comment]
Hi there.

I'm having the same problem when trying `remove `   

```
Serverless: Getting all objects in S3 bucket...
 
  Serverless Error ---------------------------------------
 
  The specified bucket does not exist
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     OS:                     darwin
     Node Version:           8.10.0
     Serverless Version:     1.40.0
```

**Comment by wviana** [comment]
BTW I don't want to delete the CF and re-deploy it. I would like to just remove it all.

**Comment by wviana** [comment]
Against my will I did deploy it againt. It deployed allright. 
But now to delete it, just got the same error: `Serverless: Getting all objects in S3 bucket...`

**Comment by alexreg** [comment]
Any updates? This is a much-needed feature for a number of us, it sounds like.

**Comment by jackfarrant** [comment]
Here is a workaround that's been working for me without needing --force:

1. Run `sls deploy` to generate an error message like: 
_Serverless Error ---------------------------------------
The serverless deployment bucket "myapp-#########" does not exist. Create it manually if you want to reuse the CloudFormation stack "myapp", or delete the stack if it is no longer required._
1. Manually create an S3 bucket called "myapp-#########" - all settings default are fine.
1. Run `sls remove` (which should now complete successfully)

Will be nice to have a proper fix soon.. =)

**Comment by lasserfox** [comment]
vaya puta mierda de sls y plugins

**Comment by dorelljames** [comment]
> Here is a workaround that's been working for me without needing --force:
> 
>     1. Run `sls deploy` to generate an error message like:
>        _Serverless Error ---------------------------------------
>        The serverless deployment bucket "myapp-#########" does not exist. Create it manually if you want to reuse the CloudFormation stack "myapp", or delete the stack if it is no longer required._
> 
>     2. Manually create an S3 bucket called "myapp-#########" - all settings default are fine.
> 
>     3. Run `sls remove` (which should now complete successfully)
> 
> 
> Will be nice to have a proper fix soon.. =)

Works well. thanks :)

**Comment by ryudice** [comment]
I'm just encountering this. How hard can it be to address this that it has been opened for 3 years?

**Comment by zaidakhterr** [comment]
> Here is a workaround that's been working for me without needing --force:
> 
> 1. Run `sls deploy` to generate an error message like:
>    _Serverless Error ---------------------------------------
>    The serverless deployment bucket "myapp-#########" does not exist. Create it manually if you want to reuse the CloudFormation stack "myapp", or delete the stack if it is no longer required._
> 2. Manually create an S3 bucket called "myapp-#########" - all settings default are fine.
> 3. Run `sls remove` (which should now complete successfully)
> 
> Will be nice to have a proper fix soon.. =)

Thank you this works. But a proper fix would be nice.

**Comment by MatteoGioioso** [comment]
> Here is a workaround that's been working for me without needing --force:
> 
> 1. Run `sls deploy` to generate an error message like:
>    _Serverless Error ---------------------------------------
>    The serverless deployment bucket "myapp-#########" does not exist. Create it manually if you want to reuse the CloudFormation stack "myapp", or delete the stack if it is no longer required._
> 2. Manually create an S3 bucket called "myapp-#########" - all settings default are fine.
> 3. Run `sls remove` (which should now complete successfully)
> 
> Will be nice to have a proper fix soon.. =)

For me this did not work:

```

Serverless Error ---------------------------------------
 
  The specified bucket does not exist 😅 
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           10.19.0
     Serverless Version:     1.32.0

```

**Comment by MatteoGioioso** [comment]
Ok solved using `SLS_DEBUG=*` in this way the name of the bucket appeared in the logs

**Comment by medikoo** [comment]
This issue was already partially addressed with https://github.com/serverless/serverless/pull/5714 where we improved message when deploy is attempted on a broken stack

Still stack removal part needs fixing. I've updated main description with fix proposal. PR is welcome!

**Comment by qbiqing** [comment]
> Here is a workaround that's been working for me without needing --force:
> 
> 1. Run `sls deploy` to generate an error message like:
>    _Serverless Error ---------------------------------------
>    The serverless deployment bucket "myapp-#########" does not exist. Create it manually if you want to reuse the CloudFormation stack "myapp", or delete the stack if it is no longer required._
> 2. Manually create an S3 bucket called "myapp-#########" - all settings default are fine.
> 3. Run `sls remove` (which should now complete successfully)
> 
> Will be nice to have a proper fix soon.. =)

When I tried creating a bucket on AWS console, I get this error:
```
An unexpected error occurred.
A conflicting conditional operation is currently in progress against this resource. Please try again.
```

**Comment by dextermb** [comment]
> Moreover, you can try to go to S3 and look for `ServerlessDeploymentBucket` and delete that too.

When looking into the issue I am having it seems like the bucket does exist, but the stack is failing to delete because of bucket related issues:

```json
{
  "StackId": "arn:aws:cloudformation:eu-west-2:<...>:stack/<...>/<...>",
  "StackName": "<...>",
  "TemplateDescription": "The AWS CloudFormation template for this Serverless application",
  "CreationTime": "2021-01-04T11:10:55.057000+00:00",
  "DeletionTime": "2021-01-04T16:21:33.926000+00:00",
  "StackStatus": "DELETE_FAILED",
  "StackStatusReason": "The following resource(s) failed to delete: [ServerlessDeploymentBucket]. ",
  "DriftInformation": {
      "StackDriftStatus": "NOT_CHECKED"
  }
}
```

When I end up looking at our S3 we have a bucket that is of a similar name to that mentioned in the `StackStatusReason`:

```
<...>-d-serverlessdeploymentbuck-<...>
```

But even then I cannot delete it as it's not empty. Should serverless be removing the contents before attempting to delete the stack?

And is it safe for me to be deleting the bucket and it's contents?

**Comment by Fomin2402** [comment]
same issue

**Comment by SherinBloemendaal** [comment]
Same issue, because another error occurred the first `sls remove` failed. Now i can't do anything anymore.

**Comment by joe4dev** [comment]
> Any progress here? I ran into this issue and was really surprised to see that you can't re-run the "remove" command once the system is in a broken state. This is a violation of [idempodency](https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation) and should definitely be fixed. A force flag is only one way around this issue -- in my opinion if the S3 bucket doesn't exist, the program should move on and attempt to tear down the other resources.

I fully agree with @gravitylow that this idempotency violation should be fixed.

There are many ways a previous `remove` operation could fail after having removed an S3 bucket and then manual cleanup is required. Image how hard it is to tear down a VPC when an AWS Firewall continuously (re-)creates a dependent security group and `remove` needs to be re-tried (after externally deleting the security group dependency).

There are also [27 potentially related issues](https://github.com/serverless/serverless/issues?q=is%3Aissue+%22The+specified+bucket+does+not+exist%22+) with the same error message "The specified bucket does not exist" and many more discussions in the Serverless forum and on StackOverflow.

**Comment by medikoo** [comment]
Thanks for all the comments. Note the `help wanted` label. PR's welcome, and implementation proposal is included in issue description

**Comment by torressam333** [comment]
TLDR: Manually delete the cloud formation stack  in AWS console, and then redeploy.

**Comment by Exy63** [comment]
I have an app create by Serverless framework and after some manipulations I couldn't delete a stack in Cloudformation which was in `UPDATE_COMPLETE_CLEANUP_IN_PROGRESS` state 
`sls deploy` and `sls remove` didn't work either.

Here is my solution: 
I just deleted `.serverless` folder from my app and tried `sls deploy` and `sls remove` 🚀

**Comment by jeremy-brooks** [comment]
In my case, I am standing up a temporary environment fresh for development and then tearing it down once work is complete. I have had multiple issues, all of which are a result of lambdas configured in serverless framework syntax depending on resources such as roles and S3 buckets configured in Cloud Formation syntax in the same stack. When the stack comes to deploy or remove, then a race condition exists between these resources, leading to random failures. One such impact is the message about a missing bucket when trying to remove a custom resource based on lambda permissions whilst trying to destroy a stack.

So the key take away for me is that there is no current way to mark resources configured in serverless as depending on another resource, which native Cloud Formation supports, and would resolve all these race condition related issues. Which is unfortunately pushing me toward a different framework.
